### PR TITLE
#921 Allow reading body multiple times

### PR DESCRIPTION
**Pull request by andriy-dmytruk**
The proposed approach to do the same as in  https://github.com/micronaut-projects/micronaut-gcp/pull/1106 did not exactly work. This is because the `InputEvent` does not allow accessing the `InputStream` directly. Instead it has a method `consumeBody(Function<InputStream>)` that should be called with implementation similar to this:
```java
<T> T consumeBody(Function<InputStream, T> function) {
      /* verify that not null and not opened twice ... */
      try {
          return function.call(this.body);
      } finally {
          this.body.close();
      }
      /* ... */
}
```
Because it closes the stream we cannot read the stream gradually and split it.

The solution is to read the stream as a whole into a byte array and then provide it to all the consumers. This will probably result in the same memory usage as in https://github.com/micronaut-projects/micronaut-gcp/pull/1106/files (since there we split each time and end up with one unused ByteBody caching the whole request). However, it will require reading the whole body before it is consumed (e.g. JSON parser cannot begin parsing the stream before whole body is read into memory).

### CODE REVIEW

**Comment by andriy-dmytruk** [comment]
It seems like with this fix: https://github.com/micronaut-projects/micronaut-servlet/pull/740 we can have an implementation using InputStreamByteBody. Not sure if this is better now.

**Comment by yawkat** [comment]
@andriy-dmytruk PTAL. My byteBody approach is detailed in this comment: https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/923/commits/fd99aaa5dbe42a5fe367c9c44c1aedb05721f3b7#diff-223b98dc29906194b0c41eeb83d19e55664789c8f55eca639a20b3e9c6ccdf51R114-R132

File: oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/HttpFunction.java [hunk 1]
```diff
@@ -101,19 +111,48 @@ protected void setupGateway(@NonNull RuntimeContext ctx) {
...
-            return response.getNativeResponse();
-        }
+        /*
+         This is a bit tricky. fnproject only allows access to the body InputStream through this
+         consumeBody method, which can only be called once and, after the lambda finishes, closes
+         the stream. This is incompatible with the ByteBody architecture.
+
+         To work around this limitation, we do a single big consumeBody around the entire request
+         processing. This hopefully encompasses most users of the body stream. Any read operations
+         inside this block are simply forwarded upstream.
```
Comment (andriy-dmytruk) [inline oraclecloud-function-http/src/main/java/io/micronaut/oraclecloud/function/http/HttpFunction.java:121]: Cool, this is great!

**Comment by andriy-dmytruk** [comment]
Rebased on latest 4.1.x branch

**Review by yawkat — APPROVED** [review]

**Comment by yawkat** [comment]
@graemerocher please review and merge

**Comment by graemerocher** [comment]
There are test failures

**Comment by yawkat** [comment]
seems there's still an issue with the FnBodyBinder, but that can be fixed in another pr. i've disabled the tck test again

**Comment by andriy-dmytruk** [comment]
@graemerocher Can we merge?

**Pull request merged**